tartufo/config.py

```diff
@@ -195,22 +196,34 @@ def configure_regexes(
     return rules
 
 
-def load_rules_from_file(rules_file: TextIO) -> Dict[str, Pattern]:
+def load_rules_from_file(rules_file: TextIO) -> Dict[str, Rule]:
     """Load a set of JSON rules from a file and return them as compiled patterns.
 
     :param rules_file: An open file handle containing a JSON dictionary of regexes
     :raises ValueError: If the rules contain invalid JSON
     """
-    regexes = {}
+    rules: Dict[str, Rule] = {}
     try:
         new_rules = json.load(rules_file)
     except json.JSONDecodeError as exc:
         raise ValueError(
             "Error loading rules from file: {}".format(rules_file.name)
         ) from exc
-    for rule in new_rules:
-        regexes[rule] = re.compile(new_rules[rule])
-    return regexes
+    for rule_name in new_rules:
+        rule_definition = new_rules[rule_name]
+        if isinstance(rule_definition, collections.Mapping):
```

Comment:
In Python it's generally considered more idiomatic, or "Pythonic", to follow an [EAFP](https://docs.python.org/3.5/glossary.html#term-eafp) (Easier to Ask for Forgiveness than Permission) style as opposed to this [LBYL](https://docs.python.org/3.5/glossary.html#term-lbyl) (Look Before You Leap) style. One major reason being that by checking for one specific type, even a base type like this, you defeat Python's duck typing.

A more idiomatic way to do this check would be something like:

```python
try:
    path_pattern = rule_definition.get("path_pattern", None)
    # ... create your rule
except AttributeError:  # The "get" will raise an AttributeError if this is not a dictionary
    # ... create your rule
```

Reply:
As a side note, this also eliminates the need to import `collections`, since the module isn't used elsewhere in here.

Reply:
In most languages, business logic by Exception is frowned on as it's a more costly operation. I try to reserve Exception handling to unexpected or edge-case scenarios.

Reply:
I am in favor of keeping the project as Pythonic as possible. This will also contributors with Python backgrounds to more easily grok and contribute to the code.

Reply:
A key here is, "In most languages". What is considered idiomatic and performant in other languages is not necessarily going to hold true for Python. The reverse is true as well. In Python, the EAFP style of coding is common and encouraged. This is in part because exception handling is not as much of a performance issue. Instead it's a mechanism you're meant to use frequently.

Reply:
In that case, you might want to submit an issue with the following repos:

- https://github.com/psf/requests/search?q=isinstance
- https://github.com/urllib3/urllib3/search?q=isinstance

Only the first two packages I looked at that are using `isinstance` before operating on an object.

In areas where they use AttributeError, they are using it for backward compatibility (if they change the methods on an object) on an object already verified by `isinstance` (as in https://github.com/psf/requests/blob/75bdc998e2d430a35d869b2abf1779bd0d34890e/requests/cookies.py#L539).